### PR TITLE
Improves efficiency of File.Exists and directory.exists

### LINKED ISSUE #20876: Directory and file exists methods could be more efficient
Related to #19717

Exists methods have a try-catch for error cases and they normalize, which is not necessary (at least on Windows). As all error cases return false, we can remove most of this overhead.

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/File.cs#L181-L210

The only pre-emptive check we might want to keep from GetFullPath is to return false if the passed in path contains an embedded null.

Maybe have a `FileSystem.FileExistsFast`? This is the current Windows one:

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/File.cs#L181-L210

Unix is already in much better state. I assume that crappy and partial (relative) paths will work.

Tests for bad and relative paths are critical, need to validate existing coverage and expand if necessary.

### PR DESCRIPTION
**Pull request by Anipik**
Fixes #20876

### CODE REVIEW

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 1]
```diff
@@ -63,24 +63,18 @@ internal static string EnsureTrailingDirectorySeparator(string fullPath)
         //
         public static bool Exists(string path)
         {
-            try
+            if (string.IsNullOrEmpty(path) || path.Contains('\0'))
+                return false;
+            
+            if (!path.StartsWith(@"\\?\") && !path.StartsWith(@"\??\"))
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: Isn't all of this Windows-specific?

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 2]
```diff
@@ -63,24 +63,18 @@ internal static string EnsureTrailingDirectorySeparator(string fullPath)
         //
         public static bool Exists(string path)
         {
-            try
+            if (string.IsNullOrEmpty(path) || path.Contains('\0'))
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: Why are we checking for `'\0'` here?
a) Isn't that a valid (though strange) character to use on Unix?
b) This is such a rare case, it doesn't seem like we should be doing a fast-path check for it.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 3]
```diff
@@ -63,24 +63,18 @@ internal static string EnsureTrailingDirectorySeparator(string fullPath)
...
-                    return false;
-
-                string fullPath = Path.GetFullPath(path);
-
-                return FileSystem.DirectoryExists(fullPath);
+				if ( path.Contains('\\') && path.Substring(path.LastIndexOf('\\')) == "\\..")
+				{
+					path = Path.GetDirectoryName(Path.GetDirectoryName(path));
+				}
+                path = path.TrimEnd(new char[] { '.', ' ' });
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: Wherever this logic ends up living, we shouldn't be allocating a new array like this each time.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs [hunk 4]
```diff
@@ -63,24 +63,18 @@ internal static string EnsureTrailingDirectorySeparator(string fullPath)
...
-                    return false;
-                if (path.Length == 0)
-                    return false;
-
-                string fullPath = Path.GetFullPath(path);
-
-                return FileSystem.DirectoryExists(fullPath);
+				if ( path.Contains('\\') && path.Substring(path.LastIndexOf('\\')) == "\\..")
+				{
+					path = Path.GetDirectoryName(Path.GetDirectoryName(path));
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/Directory.cs]: Can we use the span-based overloads to avoid intermediate string allocations?

**Comment by stephentoub** [comment]
> Improves efficiency

Can you explain how this improves the efficiency?  What is it optimizing?  Perf tests?

File: src/System.IO.FileSystem/src/System/IO/File.cs [hunk 5]
```diff
@@ -171,32 +171,16 @@ public static void Delete(string path)
...
+                return false;
 
-                return FileSystem.FileExists(path);
-            }
-            catch (ArgumentException) { }
-            catch (NotSupportedException) { } // Security can throw this on ":"
-            catch (SecurityException) { }
-            catch (IOException) { }
-            catch (UnauthorizedAccessException) { }
+            if (!path.StartsWith(@"\\?\") && !path.StartsWith(@"\??\"))
```
Comment (jkotas) [inline src/System.IO.FileSystem/src/System/IO/File.cs]: StartsWith is using CurrentCulture for comparison. Should this use Ordinal or Invariant culture instead?

File: src/System.IO.FileSystem/src/System/IO/Directory.cs (hunk 1 above)
Comment (Anipik) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Yeah it is just windows specific

Comment (stephentoub) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: Ok, thanks, then it shouldn't be in this shared portion of the implementation.

File: src/System.IO.FileSystem/src/System/IO/Directory.cs (hunk 2 above)
Comment (Anipik) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: In the previous version also, we were throwing the argument exception if the path contains an embedded null

Comment (Anipik) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/Path.Unix.cs#L26

Comment (stephentoub) [reply src/System.IO.FileSystem/src/System/IO/Directory.cs]: > we were throwing the argument exception if the path contains an embedded null

Ok

File: src/System.IO.FileSystem/src/System/IO/File.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/src/System/IO/File.cs]: In this case, Ordinal

**Comment by Anipik** [comment]
After talking offline with @JeremyKuhne. We are investigating some other approaches to do it.

**Pull request closed without merging**